docs/proposals/20190919-machinepool-api.md

```diff
@@ -0,0 +1,362 @@
+of commonalities with the MachineDeployment type. Allowing users of CAPI to leverage the
+optimizations exposed by each infrastructure provider could prove beneficial.
+  
+**Potential benefits include:**
+- Faster machine provisioning
+- Improved provisioning success rates
+- Automatic distribution of machines across availability zones if supported by the infrastructure
+  provider
+- CAPI initiated rolling update of machines
+- Higher maximum machines in a cluster (Azure limitations)
+- Auto-scaling
+
+### Goals
+
+- To expose the MachinePool API for infrastructure providers to leverage their optimizations around managing large sets of machines.
+- Support for user initiated scale up/down.
+- Support for declarative rolling update.
+
+### Non-goals/Future Work
+
+- To support enabling infrastructure provider specific autoscalers. (at least in v1alpha3)
+- To support cordon/drain during infrastructure provider specific rolling update.
+- To manage control plane nodes with the MachinePool API.
+- To integrate MachinePool with the kubernetes cluster autoscaler.
+
+## Proposal
+
+This proposal introduces the MachinePool API for the purpose of delegating the management of pools
+of machines to infrastructure provider supplied controllers.
+
+### User Stories
+
+- As an infrastructure provider author, I would like to build a controller to manage multiple
+  machines with a common configuration using my provider specific resource for doing so.
+- As a cluster operator, I would like to use MachinePool, similar to how I'm using MachineDeployment
+  today, to manage a set of machines with a common configuration.
+
+### Implementation Details/Notes/Constraints
+
+#### Data Model Changes
+
+MachinePool Spec and Status introduces the integration point for delegating the management of a set
+of machines to the infrastructure provider. Many of the fields are shared with MachineDeployment due
+to infrastructure provider's desire to enable the management of a set of machines with a single
+configuration.
+
+``` go
+type MachinePoolSpec struct
+```
+
+- **To add**
+  - **ClusterName [required]**
+    - Type: `string`
+    - Description: Name of the Cluster this machine pool belongs to.
+  - **Replicas [optional]**
+    - Type: `*int32`
+    - Description: Number of desired machine instances. Defaults to 1.
+  - **Template [required]**
+    - Type: `MachineTemplateSpec`
+    - Description: Machine Template that describes the configuration of each machine instance in a
+      machine pool.
+  - **Strategy [optional]**
+    - Type: `*MachineDeploymentStrategy`
+    - Description: Strategy to employ in replacing existing machine instances in a machine pool.
+  - **MinReadySeconds [optional]**
+    - Type: `*int32`
+    - Description: Minimum number of seconds for which a newly created machine should be ready.
+  - **ProviderIDs [optional]**
+    - Type: `[]Instance`
```

Comment:
Type string ?

Reply:
Can you explain a bit about the reason to have ProviderIDs under spec instead of under status?

Reply:
We usually put things like provider IDs in spec because the Kubernetes apiserver typically does not allow you to create a resource with status filled in. This is important if you're trying to handle a recovery situation, or any time you want to move your resources from one cluster to another. It's extremely important for `Machine` resources. If we aren't able to restore the provider ID, then we risk creating a second instance instead of re-associating with the original one.